src/main/java/org/testng/internal/ClassBasedParallelParallelWorker.java

```diff
@@ -0,0 +1,136 @@
+package org.testng.internal;
+
+import org.testng.IMethodInstance;
+import org.testng.ITestClass;
+import org.testng.ITestNGMethod;
+import org.testng.collections.Lists;
+import org.testng.collections.Maps;
+import org.testng.collections.Sets;
+import org.testng.internal.thread.graph.IWorker;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlTest;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+class ClassBasedParallelParallelWorker extends AbstractParallelWorker {
+
+    @Override
+    public List<IWorker<ITestNGMethod>> createWorkers(Arguments arguments) {
+        List<IWorker<ITestNGMethod>> result = Lists.newArrayList();
+        // Methods that belong to classes with a sequential=true or parallel=classes
+        // attribute must all be run in the same worker
+        Set<Class> sequentialClasses = Sets.newHashSet();
+        for (ITestNGMethod m : arguments.getMethods()) {
+            Class<? extends ITestClass> cls = m.getRealClass();
+            org.testng.annotations.ITestAnnotation test =
+                    arguments.getFinder().findAnnotation(cls, org.testng.annotations.ITestAnnotation.class);
+
+            // If either sequential=true or parallel=classes, mark this class sequential
+            if (isSequential(test, arguments.getTestContext().getCurrentXmlTest())) {
+                sequentialClasses.add(cls);
+            }
+        }
+
+        List<IMethodInstance> methodInstances = Lists.newArrayList();
+        for (ITestNGMethod tm : arguments.getMethods()) {
+            methodInstances.addAll(methodsToMultipleMethodInstances(tm));
+        }
+
+        Set<Class<?>> processedClasses = Sets.newHashSet();
+        Map<String, String> params = null;
+        Class<?> prevClass = null;
+        for (IMethodInstance im : methodInstances) {
+            Class<?> c = im.getMethod().getTestClass().getRealClass();
+            if (!c.equals(prevClass)) {
+                //Calculate the parameters to be injected only once per Class and NOT for every iteration.
+                params = getParameters(im);
+                prevClass = c;
+            }
+            if (sequentialClasses.contains(c)) {
+                if (!processedClasses.contains(c)) {
+                    processedClasses.add(c);
+                    if (System.getProperty("experimental") != null) {
```

Comment:
I haven't used it. So I would like to first use it, and see what it has to offer. If it's useful, maybe we can enrich the documentation etc., for it so that people can use it. If its not useful, I will remove it as part of a new PR. I have raised an issue which will capture this https://github.com/cbeust/testng/issues/1605

Reply:
👍